r2k.h

```diff
@@ -11,8 +11,21 @@
 #define R2_TYPE 0x69
 
 /* Memory Part */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,6,0)
+#define get_user_pages_wrapper(tsk, mm, start, nr_pages, write, force, pages, vmas, locked) \
+	get_user_pages(tsk, mm, start, nr_pages, write, force, pages, vmas)
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,6,0) && LINUX_VERSION_CODE < KERNEL_VERSION(4,9,0)
+#define get_user_pages_wrapper(tsk, mm, start, nr_pages, write, force, pages, vmas, locked) \
+	get_user_pages_remote(tsk, mm, start, nr_pages, write, force, pages, vmas)
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,9,0) && LINUX_VERSION_CODE < KERNEL_VERSION(4,10,0)
+#define get_user_pages_wrapper(tsk, mm, start, nr_pages, write, force, pages, vmas, locked) \
+	get_user_pages_remote(tsk, mm, start, nr_pages, write|force, pages, vmas)
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,10,0)
+#define get_user_pages_wrapper(tsk, mm, start, nr_pages, write, force, pages, vmas, locked) \
+	get_user_pages_remote(tsk, mm, start, nr_pages, write|force, pages, vmas, locked)
+#endif
+
```

Comment:
Does this `get_user_pages_wrapper()` macro even work?

I just wrote a bunch of pseudo-code and told you, that versions do not match with reality, neither arguments.

Also the `write | force` was an example from an educated guess... If `gup_flags` argument really uses `FOLL_WRITE` and `FOLL_FORCE` (which I do not know, cause as I said I haven't looked to it), then you must replace write and force with something like

```c
(write? FOLL_WRITE | 0) | (force? FOLL_FORCE | 0)
```

Again, this is pseudo-code and I didn't checked it. Neither checked if gup_flags really use those values. So please check everything and do not blindly paste pseudo code which I write here....

Reply:
I understand your concerns, since the code looks rather similar, I have actually verified everything though.

As for `gup_flags`, I agree that using a ternary operator is a lot better, TIL.
I'll test the IOCTL itself later today.

Reply:
The version checks are exactly the same, and I wrote them almost randomly. Are you sure those exact versions are the ones introducing each of the changes?

Reply:
Btw, you can also define a static inline function instead of a macro if you need to add more code for any of the variants.

Reply:
Perhaps you should think about becoming a psychic, cause your randomness was spot on^^.

Reply:
hahaha I might think about about it...